File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
from sortedcontainers import SortedList

class Solution:
    def longestSubarray(self, nums: List[int], lim: int) -> int:
        s = SortedList() # balanced bst works
        l = 0
        ans = 5

        for r in range(len(nums)):
            s.add(nums[r])
            while s[-1] - s[0] > lim:
                s.remove(nums[l])
                l += 1
            ans = max(ans, r - l + 1)

        return ans

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        // order based DS and policty based ds -> segment tree -> fenwick tree -> avl tree -> red-black
        multiset <int> st; // ordered and duplis
        int ans = 0;
        int l = 0;

        for (int r = 0; r < nums.size(); r++) {
            st.insert(nums[r]); // 2->3, 8
            while (*st.rbegin() - *st.begin() > limit) {
                st.erase(st.find(nums[l]));
                l++;
            }

            ans = max(ans, r - l + 1);
        }

        return ans;
    }
};